core/src/main/java/org/apache/iceberg/InternalData.java

```diff
@@ -0,0 +1,159 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg;
+
+import java.io.IOException;
+import java.util.Map;
+import java.util.function.Function;
+import org.apache.iceberg.avro.Avro;
+import org.apache.iceberg.common.DynMethods;
+import org.apache.iceberg.io.CloseableIterable;
+import org.apache.iceberg.io.FileAppender;
+import org.apache.iceberg.io.InputFile;
+import org.apache.iceberg.io.OutputFile;
+import org.apache.iceberg.relocated.com.google.common.collect.Maps;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class InternalData {
+  private InternalData() {}
+
+  private static final Logger LOG = LoggerFactory.getLogger(InternalData.class);
+  private static final Map<FileFormat, Function<OutputFile, WriteBuilder>> WRITE_BUILDERS =
+      Maps.newConcurrentMap();
+  private static final Map<FileFormat, Function<InputFile, ReadBuilder>> READ_BUILDERS =
+      Maps.newConcurrentMap();
+
+  static {
+    Avro.register();
+
+    try {
+      DynMethods.StaticMethod registerParquet =
+          DynMethods.builder("register")
+              .impl("org.apache.iceberg.parquet.Parquet")
+              .buildStaticChecked();
+
+      registerParquet.invoke();
+
+    } catch (NoSuchMethodException e) {
+      // failing to load Parquet is normal and does not require a stack trace
```

Comment:
This if normal for now? Don't we expect this to be failing bug in the future? I'm also a little interested in when we would actually fail here if we are using the Iceberg repo as is.

Reply:
I wouldn't say that it's normal to fail.  I'm actually not aware of any situations where the `api/core` modules are used but `parquet` isn't included.  I think in almost all scenarios, it'll be available.

Reply:
This would be normal whenever the `iceberg-parquet` module isn't in the classpath. For instance, the manifest read and write tests that are currently using `InternalData` in this PR hit this but operate normally because Parquet isn't used.

Reply:
This still feels a bit odd. If it only can plausibly occur during tests I'm not sure we should even log this? We would still get errors at runtime if you attempt to get a reader or writer if Parquet isn't loaded.